### PR TITLE
Leaderboard and Room/Round Circular Dependency 

### PR DESCRIPTION
**Pull request by Anphs**
- Solved room/round circular dependency by injecting RoundService into RoomService and abstracting RoundService from the end user's point of view.
-  Added leaderboard endpoint backed by Redis cache

### CODE REVIEW

File: central-service/services/round.go [hunk 1]
```diff
@@ -53,29 +49,10 @@ func (service *RoundService) SetDBConnection(db *gorm.DB) {
...
-		log.Printf("Error checking round limit: %v\n", err)
-		return nil, err
-	}
-	if roundLimitExceeded {
-		return nil, &RoundServiceError{
-			StatusCode: 400,
-			Message:    "Round limit exceeded",
-		}
-	}
+func (service *RoundService) CreateRound(params *RoundCreationParameters, roomID *uuid.UUID) (*models.Round, error) {
```
Comment (DedsecKnight) [inline central-service/services/round.go:52]: Thoughts on making `roomID` a part of the `RoundCreationParameters` struct?

File: central-service/services/round.go [hunk 2]
```diff
@@ -224,11 +209,224 @@ func (service *RoundService) InitiateRoundStart(roundID uint) (*time.Time, error
...
+	}
+	roundID, err := strconv.ParseUint(roundIDStr, 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	return service.getLeaderboardByRoundID(uint(roundID))
+}
+
+// First 10 bits will represent the user score and the remaining 24 bits represent the user's submission timestamp
+func compressScoreAndTimeStamp(score uint64, timestamp time.Time) float64 {
```
Comment (DedsecKnight) [inline central-service/services/round.go]: Any reason why we have to use `float64` instead of `int64`?

File: central-service/services/room.go [hunk 3]
```diff
@@ -146,46 +216,108 @@ func (service *RoomService) IsRoomAdmin(roomID string, userID string) (bool, err
...
+			StatusCode: 400,
+			Message:    "roomName must be under 32 characters in length",
+		}
+	}
+	return nil
 }
 
-type RoomServiceError struct {
-	Message string
+func (service *RoomService) CreateRound(params *RoundCreationParameters, roomID string) (*models.Round, error) {
```
Comment (DedsecKnight) [inline central-service/services/room.go:259]: Thoughts on making `roomID` part of `RoundCreationParameters` struct?

File: central-service/services/round.go (hunk 2 above)
Comment (Anphs) [reply central-service/services/round.go]: Sorted set score expects a float

File: central-service/controllers/room.go [hunk 4]
```diff
@@ -78,19 +78,73 @@ func (controller *RoomController) LeaveRoomEndpoint(c echo.Context) error {
...
+}
+
 func (controller *RoomController) InitializeRoutes(g *echo.Group) {
 	g.POST("/", controller.CreateNewRoomEndpoint)
-	g.GET("/:roomID", controller.FindRoomEndpoint)
 	g.POST("/:roomID/join", controller.JoinRoomEndpoint)
 	g.POST("/:roomID/leave", controller.LeaveRoomEndpoint)
-}
\ No newline at end of file
+	g.POST("/:roomID/new-round", controller.CreateNewRoundEndpoint)
```
Comment (DedsecKnight) [inline central-service/controllers/room.go]: nit: can we get rid of the `-` somehow? Like rewrite the endpoint as something like `/:roomID/rounds/create`

**Review by DedsecKnight — APPROVED** [review]
:100:

**Review by spierceVR — APPROVED** [review]
looks good

**Pull request merged**